File: generated_code/GPT4/Java/068_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;
        Node(int key) {
            this.key = key;
        }
    }

    private Node root;

    // Public insert method
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }
        Node z = root, p = null;
        while (z != null) {
            p = z;
            if (key < z.key) z = z.left;
            else if (key > z.key) z = z.right;
            else {
                splay(z);
                return; // key already exists, do nothing
            }
        }
        Node n = new Node(key);
        n.parent = p;
        if (key < p.key) p.left = n;
        else p.right = n;
        splay(n);
    }

    // Public delete method
    public void delete(int key) {
        Node node = find(root, key);
        if (node == null) return;
        splay(node);
        if (node.left != null) {
            Node maxLeft = subtreeMax(node.left);
            splay(maxLeft);
            maxLeft.right = node.right;
            if (node.right != null) node.right.parent = maxLeft;
            root = maxLeft;
            maxLeft.parent = null;
        } else if (node.right != null) {
            root = node.right;
            root.parent = null;
        } else {
            root = null;
        }
    }

    // Public search method
    public Integer search(int key) {
        Node node = find(root, key);
        if (node != null) {
            splay(node);
            return node.key;
        }
        return null;
    }

    // Helper: find node with given key
    private Node find(Node curr, int key) {
        while (curr != null) {
            if (key < curr.key) curr = curr.left;
            else if (key > curr.key) curr = curr.right;
            else return curr;
        }
        return null;
    }

    // Helper: splay node to root
    private void splay(Node x) {
        if (x == null) return;
        while (x.parent != null) {
            Node p = x.parent;
            Node g = p.parent;
            if (g == null) {
                // Zig
                if (x == p.left) rightRotate(p);
                else leftRotate(p);
            } else if (x == p.left && p == g.left) {
                // Zig-Zig
                rightRotate(g);
                rightRotate(p);
            } else if (x == p.right && p == g.right) {
                // Zig-Zig
                leftRotate(g);
                leftRotate(p);
            } else if (x == p.right && p == g.left) {
                // Zig-Zag
                leftRotate(p);
                rightRotate(g);
            } else if (x == p.left && p == g.right) {
                // Zig-Zag
                rightRotate(p);
                leftRotate(g);
            }
        }
        root = x;
    }

    // Helper: right rotate
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y == null) return;
        x.left = y.right;
        if (y.right != null) y.right.parent = x;
        y.parent = x.parent;
        if (x.parent == null) root = y;
        else if (x == x.parent.left) x.parent.left = y;
        else x.parent.right = y;
        y.right = x;
        x.parent = y;
    }

    // Helper: left rotate
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y == null) return;
        x.right = y.left;
        if (y.left != null) y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null) root = y;
        else if (x == x.parent.left) x.parent.left = y;
        else x.parent.right = y;
        y.left = x;
        x.parent = y;
    }

    // Helper: subtree maximum
    private Node subtreeMax(Node x) {
        while (x.right != null) x = x.right;
        return x;
    }

    // Optional: for testing
    public void inorder() {
        inorder(root);
        System.out.println();
    }

    private void inorder(Node n) {
        if (n == null) return;
        inorder(n.left);
        System.out.print(n.key + " ");
        inorder(n.right);
    }

    // Main for demonstration (not required)
    public static void main(String[] args) {
        SplayTree st = new SplayTree();
        st.insert(10);
        st.insert(20);
        st.insert(5);
        st.insert(8);
        st.insert(15);
        st.inorder(); // 5 8 10 15 20

        System.out.println(st.search(15)); // 15
        st.inorder(); // 5 8 10 15 20 (but 15 is root internally)

        st.delete(10);
        st.inorder(); // 5 8 15 20
        System.out.println(st.search(100)); // null
    }
}

File: generated_code/GPT4/Python/068_SplayTree.py
class SplayTree:
    class Node:
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    # --- Rotation helpers ---
    def _rotate_left(self, x):
        y = x.right
        if not y:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, x):
        y = x.left
        if not y:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.right = x
        x.parent = y

    # --- Splaying ---
    def _splay(self, x):
        if not x:
            return
        while x.parent:
            if not x.parent.parent:
                # Zig step
                if x.parent.left == x:
                    self._rotate_right(x.parent)
                else:
                    self._rotate_left(x.parent)
            else:
                p = x.parent
                g = p.parent
                if g.left == p and p.left == x:
                    # Zig-Zig step (left-left)
                    self._rotate_right(g)
                    self._rotate_right(p)
                elif g.right == p and p.right == x:
                    # Zig-Zig step (right-right)
                    self._rotate_left(g)
                    self._rotate_left(p)
                elif g.left == p and p.right == x:
                    # Zig-Zag step (left-right)
                    self._rotate_left(p)
                    self._rotate_right(g)
                elif g.right == p and p.left == x:
                    # Zig-Zag step (right-left)
                    self._rotate_right(p)
                    self._rotate_left(g)

    # --- Search ---
    def search(self, key):
        node = self.root
        parent = None
        while node:
            if key == node.key:
                self._splay(node)
                return True
            parent = node
            if key < node.key:
                node = node.left
            else:
                node = node.right
        if parent:
            self._splay(parent)
        return False

    # --- Insert ---
    def insert(self, key):
        if not self.root:
            self.root = self.Node(key)
            return
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                self._splay(node)
                return  # Key already in set, do nothing
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        new_node = self.Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        self._splay(new_node)

    # --- Delete ---
    def delete(self, key):
        if not self.root:
            return
        # Search and splay the node (or its parent if not found) to root
        node = self.root
        parent = None
        while node:
            if key == node.key:
                self._splay(node)
                break
            parent = node
            if key < node.key:
                node = node.left
            else:
                node = node.right
        else:
            if parent:
                self._splay(parent)
            return  # Key not found, do nothing

        # Now self.root is the node to delete
        if not self.root.left:
            self._replace_root(self.root.right)
        elif not self.root.right:
            self._replace_root(self.root.left)
        else:
            # Both children exist: find maximum in left subtree
            left_sub = self.root.left
            left_sub.parent = None
            max_left = left_sub
            while max_left.right:
                max_left = max_left.right
            # Splay max_left to the root of left_sub
            self._splay_node_in_subtree(max_left, left_sub)
            # Attach right subtree
            max_left.right = self.root.right
            if self.root.right:
                self.root.right.parent = max_left
            self.root = max_left
            self.root.parent = None

    def _replace_root(self, node):
        self.root = node
        if node:
            node.parent = None

    def _splay_node_in_subtree(self, x, subtree_root):
        # Splay x to the root of the subtree rooted at subtree_root
        while x.parent:
            if x.parent == subtree_root:
                if x.parent.left == x:
                    self._rotate_right(x.parent)
                else:
                    self._rotate_left(x.parent)
                break
            p = x.parent
            g = p.parent
            if g.left == p and p.left == x:
                self._rotate_right(g)
                self._rotate_right(p)
            elif g.right == p and p.right == x:
                self._rotate_left(g)
                self._rotate_left(p)
            elif g.left == p and p.right == x:
                self._rotate_left(p)
                self._rotate_right(g)
            elif g.right == p and p.left == x:
                self._rotate_right(p)
                self._rotate_left(g)

    # Optional: for debugging or testing
    def to_list(self):
        res = []
        def inorder(node):
            if node:
                inorder(node.left)
                res.append(node.key)
                inorder(node.right)
        inorder(self.root)
        return res